openflow13/nx_action.go

```diff
@@ -0,0 +1,956 @@
+		return errors.New("protocol hash should be exclusively with random")
+	}
+	a.Flags |= NX_NAT_F_PROTO_HASH
+	return nil
+}
+
+func (a *NXActionCTNAT) SetRandom() error {
+	if a.Flags&NX_NAT_F_PROTO_HASH != 0 {
+		return errors.New("random should be exclusively with protocol hash")
+	}
+	a.Flags |= NX_NAT_F_PROTO_RANDOM
+	return nil
+}
+
+func (a *NXActionCTNAT) SetPersistent() error {
+	a.Flags |= NX_NAT_F_PERSISTENT
+	return nil
+}
+
+func (a *NXActionCTNAT) SetRangeIPv4Min(ipMin net.IP) {
+	a.rangeIPv4Min = ipMin
+	a.rangePresent |= NX_NAT_RANGE_IPV4_MIN
+	a.Length += 4
+}
+func (a *NXActionCTNAT) SetRangeIPv4Max(ipMax net.IP) {
+	a.rangeIPv4Max = ipMax
+	a.rangePresent |= NX_NAT_RANGE_IPV4_MAX
+	a.Length += 4
+}
+func (a *NXActionCTNAT) SetRangeIPv6Min(ipMin net.IP) {
+	a.rangeIPv6Min = ipMin
+	a.rangePresent |= NX_NAT_RANGE_IPV6_MIN
+	a.Length += 16
+}
+func (a *NXActionCTNAT) SetRangeIPv6Max(ipMax net.IP) {
+	a.rangeIPv6Max = ipMax
+	a.rangePresent |= NX_NAT_RANGE_IPV6_MAX
+	a.Length += 16
+}
+func (a *NXActionCTNAT) SetRangeProtoMin(protoMin *uint16) {
+	a.rangeProtoMin = protoMin
+	a.rangePresent |= NX_NAT_RANGE_PROTO_MIN
+	a.Length += 2
+}
+func (a *NXActionCTNAT) SetRangeProtoMax(protoMax *uint16) {
+	a.rangeProtoMax = protoMax
+	a.rangePresent |= NX_NAT_RANGE_PROTO_MAX
+	a.Length += 2
+}
+
+func (a *NXActionCTNAT) UnmarshalBinary(data []byte) error {
+	n := 0
+	a.NXActionHeader = new(NXActionHeader)
+	err := a.NXActionHeader.UnmarshalBinary(data[n:])
+	n += int(a.NXActionHeader.Len())
+	if len(data) < int(a.Len()) {
+		return errors.New("the byte array has wrong size to unmarshal an NXActionCTNAT message")
+	}
+	copy(a.pad, data[n:n+2])
+	n += 2
+	a.Flags = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	a.rangePresent = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	if a.rangePresent&NX_NAT_RANGE_IPV4_MIN != 0 {
+		a.rangeIPv4Min = net.IPv4(data[n], data[n+1], data[n+2], data[n+3])
+		n += 4
+	}
+	if a.rangePresent&NX_NAT_RANGE_IPV4_MAX != 0 {
+		a.rangeIPv4Max = net.IPv4(data[n], data[n+1], data[n+2], data[n+3])
+		n += 4
+	}
+	if a.rangePresent&NX_NAT_RANGE_IPV6_MIN != 0 {
+		a.rangeIPv6Min = make([]byte, 16)
+		copy(a.rangeIPv6Min, data[n:n+16])
+		n += 16
+	}
+	if a.rangePresent&NX_NAT_RANGE_IPV6_MAX != 0 {
+		a.rangeIPv6Max = make([]byte, 16)
+		copy(a.rangeIPv6Max, data[n:n+16])
+		n += 16
+	}
+	if a.rangePresent&NX_NAT_RANGE_PROTO_MIN != 0 {
+		portMin := binary.BigEndian.Uint16(data[n:])
+		a.rangeProtoMin = &portMin
+		n += 2
+	}
+	if a.rangePresent&NX_NAT_RANGE_PROTO_MAX != 0 {
+		portMax := binary.BigEndian.Uint16(data[n:])
+		a.rangeProtoMin = &portMax
+		n += 2
+	}
+
+	return err
+}
+
+// nx_action_output_reg
+type NXActionOutputReg struct {
+	*NXActionHeader
+	OfsNbits uint16      // (ofs << 6 | (Nbits -1)
+	SrcField *MatchField // source nxm_nx_reg
+	MaxLen   uint16      // Max length to send to controller if chosen port is OFPP_CONTROLLER
+	zero     []uint8     // 6 uint8 with all Value as 0, reserved
```

Comment:
I think for such a case, a Go array may be simpler (you can use default initialization) and more efficient: `zeros [6]uint8`. It also makes it clear that there are 6 uint8 values.

Reply:
Done.